keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,364 @@
+  * While feature gate provides a [path to deprecate existing `--experimental` features]((#path-to-migrate-existing-experimental-features)), and we will migrate several `--experimental` during the feature gate implementation as examples, the completion of the KEP does not require migrating all existing `--experimental` features. It can be done as followup tasks.
+
+* Use feature gate as a mechanism to fix bugs.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+It would be easier for developers to add new features without worrying too much about breaking etcd, because the feature would be Alpha first and disabled by default, and new changes would be mostly gated behind the feature gate.
+
+It would be a smoother process to move the feature to the next stage without the need to introduce a potentially breaking change of an `--experimental` flag.
+
+#### Story 2
+
+Users of etcd cluster would be able to find out which features are enabled in the etcd server and decide on how to use them downstream. 
+
+Currently Kubernetes uses [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to check if a feature is supported by etcd, which basically checks the etcd version and compares the version with a hard-coded map of feature availability for different etcd versions.
+
+#### Story 3
+
+In a HA cluster, users should be able to enable/disable a feature for each server without bring down the whole cluster, by restarting the servers with the new feature gate flag one by one. 
+
+#### Story 4
+
+Users of etcd cluster would like to know if they could use a new feature, but it requires all the etcd servers to agree on the feature enablement. Until the feature gate can handle cluster level features (another KEP), we will not expose querying feature enablement of any cluster level features, because we cannot risk possible conflicting settings that might break cluster correctness.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Is feature enablement state a server level or cluster level property?
+
+There are simple features like `ExperimentalEnableDistributedTracing` (enables distributed tracing using OpenTelemetry protocol), which could work fine at the local server level.
+
+There are also features like `ExperimentalEnableLeaseCheckpoint` (enables leader to send regular checkpoints to other members to prevent reset of remaining TTL on leader change), if different server nodes have different enablement values for `ExperimentalEnableLeaseCheckpoint`, the results would be inconsistent depending on which one is the leader and confusing.
+
+There are also more critial features like changing the Apply logic of a raft message. All members have to agree on whether to use the new logic or old logic, otherwise it could could break correctness.
```

Comment:
```suggestion
There are also more critial features like changing the Apply logic of a raft message. All members have to agree on whether to use the new logic or old logic, otherwise it could break correctness.
```